cmd/skaffold/app/cmd/diagnose.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2018 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/version"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/watch"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+// NewCmdDiagnose describes the CLI command to diagnose skaffold.
+func NewCmdDiagnose(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "diagnose",
+		Short: "Run a diagnostic on Skaffold",
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return doDiagnose(out)
+		},
+	}
+	cmd.Flags().StringVarP(&opts.ConfigurationFile, "filename", "f", "skaffold.yaml", "Filename or URL to the pipeline file")
+	return cmd
+}
+
+func doDiagnose(out io.Writer) error {
+	_, config, err := newRunner(opts)
+	if err != nil {
+		return errors.Wrap(err, "creating runner")
+	}
+
+	fmt.Fprintln(out, "Skaffold version:", version.Get().GitCommit)
+	fmt.Fprintln(out, "Configuration version:", config.APIVersion)
+	fmt.Fprintln(out, "Number of artifacts:", len(config.Build.Artifacts))
+
+	fmt.Fprintln(out, "\nRunning a diagnostic on artifacts and their dependencies:")
+
+	if err := diagnoseArtifacts(out, config.Build.Artifacts); err != nil {
+		return errors.Wrap(err, "running diagnostic on artifacts")
+	}
+
+	fmt.Fprintln(out, "\nSecond run (that benefits from cache):")
+
+	if err := diagnoseArtifacts(out, config.Build.Artifacts); err != nil {
+		return errors.Wrap(err, "running second diagnostic on artifacts")
+	}
+
+	return nil
+}
+
+func diagnoseArtifacts(out io.Writer, artifacts []*latest.Artifact) error {
```

Comment:
it might also be worth trying to infer some of the builder- and deployer- specific information for each artifact. e.g. docker version for a docker artifact, bazel version for a bazel artifact, kubectl version, helm version, etc.